File: apps/ios/Riot/Release/ReleaseFixture.swift
import CryptoKit
import Foundation

public enum ReleaseFixtureError: Error, Equatable {
    case digestMismatch
    case malformedJSON
    case invalidRootKeys
    case invalidIdentifierKeys
    case invalidStateKeys
    case invalidSchemaVersion
    case invalidFixtureRevision
    case nonSyntheticKind
    case invalidFixedClock
    case invalidStateInventory
    case invalidIdentifier
    case mismatchedStateIdentifier
    case invalidNarrativeValue
    case prohibitedData
}

public struct ReleaseFixture: Sendable, Equatable, Decodable {
    public struct Identifiers: Sendable, Equatable, Decodable {
        public let communityId: String
        public let contributorId: String
        public let stateIds: [String: String]
    }

    public struct NarrativeState: Sendable, Equatable, Decodable {
        public let id: String
        public let surface: String
        public let headline: String
        public let supportingCopy: String
        public let communityId: String
        public let contributorId: String
        public let entryId: String
    }

    public let schemaVersion: Int
    public let fixtureRevision: String
    public let fixtureKind: String
    public let fixedClock: String
    public let identifiers: Identifiers
    public let narrativeStates: [NarrativeState]

    private static let pinnedDigest = "930a9c5aa06dea920b0502dbd72b6b2bf00d1b4cb9405b99e69e00d035640469"
    private static let rootKeys: Set<String> = [
        "schemaVersion", "fixtureRevision", "fixtureKind", "fixedClock",
        "identifiers", "narrativeStates",
    ]
    private static let identifierKeys: Set<String> = ["communityId", "contributorId", "stateIds"]
    private static let stateKeys: Set<String> = [
        "id", "surface", "headline", "supportingCopy",
        "communityId", "contributorId", "entryId",
    ]
    private static let expectedStates: [(id: String, surface: String, headline: String, copy: String)] = [
        ("community-newswire", "spaces-home", "Your community. Your newswire.", "Follow updates from a community you choose."),
        ("signed-publishing", "compose", "Publish signed updates from the field.", "Signatures show source and integrity, not whether a claim is true."),
        ("labels-and-signatures", "newswire", "Read signatures and community editorial labels.", "See signed source details. Community editorial labels are community signals, not independent factual verification."),
        ("community-tools", "apps-checklists", "Carry useful tools with the community.", "Open a shared checklist alongside community updates."),
        ("nearby-exchange", "nearby", "Exchange updates nearby.", "Experimental: exchange updates directly with a nearby device."),
        ("offline-copy", "offline-copy", "Keep a local copy available offline.", "Keep a local copy ready to read without a connection."),
    ]
    private static let prohibitedKeys: Set<String> = [
        "person", "personname", "name", "email", "phone", "location", "address",
        "latitude", "longitude", "coordinates", "notification", "devicetoken",
        "apnstoken", "fcmtoken", "private", "privatecommunity", "url", "hostname",
        "ipaddress", "npub", "nsec", "note",
    ]

    public static func decode(bytes: Data) throws -> ReleaseFixture {
        guard digest(bytes) == pinnedDigest else {
            throw ReleaseFixtureError.digestMismatch
        }
        return try validateSemantics(bytes: bytes)
    }

    static func validateSemantics(bytes: Data) throws -> ReleaseFixture {
        let raw: Any
        do {
            raw = try JSONSerialization.jsonObject(with: bytes)
        } catch {
            throw ReleaseFixtureError.malformedJSON
        }
        guard let root = raw as? [String: Any] else {
            throw ReleaseFixtureError.malformedJSON
        }
        try rejectProhibitedData(raw)
        guard Set(root.keys) == rootKeys else {
            throw ReleaseFixtureError.invalidRootKeys
        }
        guard let identifiersObject = root["identifiers"] as? [String: Any],
              Set(identifiersObject.keys) == identifierKeys else {
            throw ReleaseFixtureError.invalidIdentifierKeys
        }
        guard let statesObject = root["narrativeStates"] as? [[String: Any]] else {
            throw ReleaseFixtureError.malformedJSON
        }
        guard statesObject.allSatisfy({ Set($0.keys) == stateKeys }) else {
            throw ReleaseFixtureError.invalidStateKeys
        }
        let fixture: ReleaseFixture
        do {
            fixture = try JSONDecoder().decode(ReleaseFixture.self, from: bytes)
        } catch {
            throw ReleaseFixtureError.malformedJSON
        }
        try fixture.validate()
        return fixture
    }

    private func validate() throws {
        guard schemaVersion == 1 else { throw ReleaseFixtureError.invalidSchemaVersion }
        guard fixtureRevision == "riot-1.0-synthetic-v1" else {
            throw ReleaseFixtureError.invalidFixtureRevision
        }
        guard fixtureKind == "synthetic" else { throw ReleaseFixtureError.nonSyntheticKind }
        guard fixedClock == "2026-07-24T12:00:00.000Z" else {
            throw ReleaseFixtureError.invalidFixedClock
        }
        guard narrativeStates.map(\.id) == Self.expectedStates.map(\.id),
              identifiers.stateIds.keys.count == Self.expectedStates.count,
              Set(identifiers.stateIds.keys) == Set(Self.expectedStates.map(\.id)) else {
            throw ReleaseFixtureError.invalidStateInventory
        }

        try Self.requireIdentifier(
            identifiers.communityId,
            label: "riot-release-fixture:v1:community"
        )
        try Self.requireIdentifier(
            identifiers.contributorId,
            label: "riot-release-fixture:v1:contributor"
        )
        for (index, expected) in Self.expectedStates.enumerated() {
            guard let stateIdentifier = identifiers.stateIds[expected.id] else {
                throw ReleaseFixtureError.invalidStateInventory
            }
            try Self.requireIdentifier(
                stateIdentifier,
                label: "riot-release-fixture:v1:\(expected.id)"
            )
            let state = narrativeStates[index]
            guard state.surface == expected.surface,
                  state.headline == expected.headline,
                  state.supportingCopy == expected.copy else {
                throw ReleaseFixtureError.invalidNarrativeValue
            }
            guard state.communityId == identifiers.communityId,
                  state.contributorId == identifiers.contributorId,
                  state.entryId == stateIdentifier else {
                throw ReleaseFixtureError.mismatchedStateIdentifier
            }
        }
    }

    private static func requireIdentifier(_ identifier: String, label: String) throws {
        guard identifier.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil,
              identifier == digest(Data(label.utf8)) else {
            throw ReleaseFixtureError.invalidIdentifier
        }
    }

    private static func rejectProhibitedData(_ value: Any) throws {
        if let object = value as? [String: Any] {
            for (key, child) in object {
                guard !prohibitedKeys.contains(key.lowercased()) else {
                    throw ReleaseFixtureError.prohibitedData
                }
                try rejectProhibitedData(child)
            }
        } else if let array = value as? [Any] {
            for child in array {
                try rejectProhibitedData(child)
            }
        } else if let string = value as? String, isProhibitedValue(string) {
            throw ReleaseFixtureError.prohibitedData
        }
    }

    private static func isProhibitedValue(_ value: String) -> Bool {
        let lower = value.lowercased()
        if lower.contains("person: ana") ||
            lower.contains("ana@example.com") ||
            lower.contains("+64 21 555 0100") ||
            lower.contains("123 main street") ||
            lower.contains("37.7749,-122.4194") ||
            lower.contains("exponentpushtoken[fixture]") ||
            lower.contains("private community") ||
            lower.contains("https://") ||
            lower.contains("http://") ||
            lower.contains("riot.protest.net") ||
            lower.contains("203.0.113.1") {
            return true
        }
        return lower.hasPrefix("npub1") || lower.hasPrefix("nsec1") || lower.hasPrefix("note1")
    }

    private static func digest(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

File: apps/android/app/src/main/kotlin/org/riot/evidence/ReleaseFixture.kt
package org.riot.evidence

import java.security.MessageDigest
import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

enum class ReleaseFixtureContractError {
    DIGEST_MISMATCH,
    MALFORMED_JSON,
    INVALID_ROOT_KEYS,
    INVALID_IDENTIFIER_KEYS,
    INVALID_STATE_KEYS,
    INVALID_SCHEMA_VERSION,
    INVALID_FIXTURE_REVISION,
    NON_SYNTHETIC_KIND,
    INVALID_FIXED_CLOCK,
    INVALID_STATE_INVENTORY,
    INVALID_IDENTIFIER,
    MISMATCHED_STATE_IDENTIFIER,
    INVALID_NARRATIVE_VALUE,
    PROHIBITED_DATA,
}

class ReleaseFixtureContractException(
    val error: ReleaseFixtureContractError,
) : IllegalArgumentException(error.name)

data class ReleaseFixtureIdentifiers(
    val communityId: String,
    val contributorId: String,
    val stateIds: Map<String, String>,
)

data class ReleaseFixtureNarrativeState(
    val id: String,
    val surface: String,
    val headline: String,
    val supportingCopy: String,
    val communityId: String,
    val contributorId: String,
    val entryId: String,
)

data class ReleaseFixture(
    val schemaVersion: Int,
    val fixtureRevision: String,
    val fixtureKind: String,
    val fixedClock: String,
    val identifiers: ReleaseFixtureIdentifiers,
    val narrativeStates: List<ReleaseFixtureNarrativeState>,
) {
    companion object {
        private const val PINNED_DIGEST =
            "930a9c5aa06dea920b0502dbd72b6b2bf00d1b4cb9405b99e69e00d035640469"
        private val ROOT_KEYS = setOf(
            "schemaVersion", "fixtureRevision", "fixtureKind", "fixedClock",
            "identifiers", "narrativeStates",
        )
        private val IDENTIFIER_KEYS = setOf("communityId", "contributorId", "stateIds")
        private val STATE_KEYS = setOf(
            "id", "surface", "headline", "supportingCopy",
            "communityId", "contributorId", "entryId",
        )
        private val EXPECTED_STATES = listOf(
            ExpectedState(
                "community-newswire",
                "spaces-home",
                "Your community. Your newswire.",
                "Follow updates from a community you choose.",
            ),
            ExpectedState(
                "signed-publishing",
                "compose",
                "Publish signed updates from the field.",
                "Signatures show source and integrity, not whether a claim is true.",
            ),
            ExpectedState(
                "labels-and-signatures",
                "newswire",
                "Read signatures and community editorial labels.",
                "See signed source details. Community editorial labels are community signals, not independent factual verification.",
            ),
            ExpectedState(
                "community-tools",
                "apps-checklists",
                "Carry useful tools with the community.",
                "Open a shared checklist alongside community updates.",
            ),
            ExpectedState(
                "nearby-exchange",
                "nearby",
                "Exchange updates nearby.",
                "Experimental: exchange updates directly with a nearby device.",
            ),
            ExpectedState(
                "offline-copy",
                "offline-copy",
                "Keep a local copy available offline.",
                "Keep a local copy ready to read without a connection.",
            ),
        )
        private val PROHIBITED_KEYS = setOf(
            "person", "personname", "name", "email", "phone", "location", "address",
            "latitude", "longitude", "coordinates", "notification", "devicetoken",
            "apnstoken", "fcmtoken", "private", "privatecommunity", "url", "hostname",
            "ipaddress", "npub", "nsec", "note",
        )

        fun decode(bytes: ByteArray): ReleaseFixture {
            requireContract(sha256(bytes) == PINNED_DIGEST, ReleaseFixtureContractError.DIGEST_MISMATCH)
            return validateSemantics(bytes)
        }

        internal fun validateSemantics(bytes: ByteArray): ReleaseFixture {
            val tokener = JSONTokener(bytes.toString(Charsets.UTF_8))
            val parsed = try {
                tokener.nextValue()
            } catch (_: Exception) {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            if (parsed !is JSONObject) {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            // JSONTokener stops after the root object; only trailing whitespace
            // (for example the final LF) may follow, anything else is malformed.
            while (tokener.more()) {
                if (!tokener.next().isWhitespace()) {
                    fail(ReleaseFixtureContractError.MALFORMED_JSON)
                }
            }
            val root = parsed
            rejectProhibitedData(root)
            requireContract(root.keysSet() == ROOT_KEYS, ReleaseFixtureContractError.INVALID_ROOT_KEYS)

            val identifiersObject = root.optJSONObject("identifiers")
                ?: fail(ReleaseFixtureContractError.INVALID_IDENTIFIER_KEYS)
            requireContract(
                identifiersObject.keysSet() == IDENTIFIER_KEYS,
                ReleaseFixtureContractError.INVALID_IDENTIFIER_KEYS,
            )
            val statesArray = try {
                root.getJSONArray("narrativeStates")
            } catch (_: Exception) {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            repeat(statesArray.length()) { index ->
                val state = try {
                    statesArray.getJSONObject(index)
                } catch (_: Exception) {
                    fail(ReleaseFixtureContractError.MALFORMED_JSON)
                }
                requireContract(state.keysSet() == STATE_KEYS, ReleaseFixtureContractError.INVALID_STATE_KEYS)
            }

            val fixture = try {
                decodeObject(root, identifiersObject, statesArray)
            } catch (error: ReleaseFixtureContractException) {
                throw error
            } catch (_: Exception) {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            fixture.validate()
            return fixture
        }

        private fun decodeObject(
            root: JSONObject,
            identifiersObject: JSONObject,
            statesArray: JSONArray,
        ): ReleaseFixture {
            val stateIdsObject = identifiersObject.getJSONObject("stateIds")
            val stateIds = stateIdsObject.keys().asSequence().associateWith { stateIdsObject.getString(it) }
            val states = (0 until statesArray.length()).map { index ->
                statesArray.getJSONObject(index).run {
                    ReleaseFixtureNarrativeState(
                        id = getString("id"),
                        surface = getString("surface"),
                        headline = getString("headline"),
                        supportingCopy = getString("supportingCopy"),
                        communityId = getString("communityId"),
                        contributorId = getString("contributorId"),
                        entryId = getString("entryId"),
                    )
                }
            }
            return ReleaseFixture(
                schemaVersion = exactSchemaVersion(root),
                fixtureRevision = root.getString("fixtureRevision"),
                fixtureKind = root.getString("fixtureKind"),
                fixedClock = root.getString("fixedClock"),
                identifiers = ReleaseFixtureIdentifiers(
                    communityId = identifiersObject.getString("communityId"),
                    contributorId = identifiersObject.getString("contributorId"),
                    stateIds = stateIds,
                ),
                narrativeStates = states,
            )
        }

        // org.json getInt truncates 1.9 to 1; Swift JSONDecoder rejects any
        // non-integral or out-of-range schemaVersion as malformed. Match that
        // exactly: only an integral number representable as Int is accepted.
        private fun exactSchemaVersion(root: JSONObject): Int = when (val raw = root.get("schemaVersion")) {
            is Int -> raw
            is Long -> if (raw.toInt().toLong() == raw) raw.toInt() else {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            is Double -> if (raw % 1.0 == 0.0 && raw.toInt().toDouble() == raw) raw.toInt() else {
                fail(ReleaseFixtureContractError.MALFORMED_JSON)
            }
            else -> fail(ReleaseFixtureContractError.MALFORMED_JSON)
        }

        private fun ReleaseFixture.validate() {
            requireContract(schemaVersion == 1, ReleaseFixtureContractError.INVALID_SCHEMA_VERSION)
            requireContract(
                fixtureRevision == "riot-1.0-synthetic-v1",
                ReleaseFixtureContractError.INVALID_FIXTURE_REVISION,
            )
            requireContract(fixtureKind == "synthetic", ReleaseFixtureContractError.NON_SYNTHETIC_KIND)
            requireContract(
                fixedClock == "2026-07-24T12:00:00.000Z",
                ReleaseFixtureContractError.INVALID_FIXED_CLOCK,
            )
            val expectedIds = EXPECTED_STATES.map { it.id }
            requireContract(
                narrativeStates.map { it.id } == expectedIds &&
                    identifiers.stateIds.keys == expectedIds.toSet(),
                ReleaseFixtureContractError.INVALID_STATE_INVENTORY,
            )

            requireIdentifier(identifiers.communityId, "riot-release-fixture:v1:community")
            requireIdentifier(identifiers.contributorId, "riot-release-fixture:v1:contributor")
            EXPECTED_STATES.forEachIndexed { index, expected ->
                val stateIdentifier = identifiers.stateIds[expected.id]
                    ?: fail(ReleaseFixtureContractError.INVALID_STATE_INVENTORY)
                requireIdentifier(stateIdentifier, "riot-release-fixture:v1:${expected.id}")
                val state = narrativeStates[index]
                requireContract(
                    state.surface == expected.surface &&
                        state.headline == expected.headline &&
                        state.supportingCopy == expected.copy,
                    ReleaseFixtureContractError.INVALID_NARRATIVE_VALUE,
                )
                requireContract(
                    state.communityId == identifiers.communityId &&
                        state.contributorId == identifiers.contributorId &&
                        state.entryId == stateIdentifier,
                    ReleaseFixtureContractError.MISMATCHED_STATE_IDENTIFIER,
                )
            }
        }

        private fun requireIdentifier(identifier: String, label: String) {
            requireContract(
                identifier.matches(Regex("[0-9a-f]{64}")) &&
                    identifier == sha256(label.toByteArray()),
                ReleaseFixtureContractError.INVALID_IDENTIFIER,
            )
        }

        private fun rejectProhibitedData(value: Any?) {
            when (value) {
                is JSONObject -> value.keys().asSequence().forEach { key ->
                    requireContract(
                        key.lowercase() !in PROHIBITED_KEYS,
                        ReleaseFixtureContractError.PROHIBITED_DATA,
                    )
                    rejectProhibitedData(value.get(key))
                }
                is JSONArray -> repeat(value.length()) { rejectProhibitedData(value.get(it)) }
                is String -> requireContract(
                    !isProhibitedValue(value),
                    ReleaseFixtureContractError.PROHIBITED_DATA,
                )
            }
        }

        private fun isProhibitedValue(value: String): Boolean {
            val lower = value.lowercase()
            return listOf(
                "person: ana", "ana@example.com", "+64 21 555 0100", "123 main street",
                "37.7749,-122.4194", "exponentpushtoken[fixture]", "private community",
                "https://", "http://", "riot.protest.net", "203.0.113.1",
            ).any(lower::contains) ||
                lower.startsWith("npub1") ||
                lower.startsWith("nsec1") ||
                lower.startsWith("note1")
        }

        private fun JSONObject.keysSet(): Set<String> = keys().asSequence().toSet()

        private fun sha256(bytes: ByteArray): String =
            MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

        private fun requireContract(condition: Boolean, error: ReleaseFixtureContractError) {
            if (!condition) fail(error)
        }

        private fun fail(error: ReleaseFixtureContractError): Nothing =
            throw ReleaseFixtureContractException(error)
    }
}

private data class ExpectedState(
    val id: String,
    val surface: String,
    val headline: String,
    val copy: String,
)
